[WIP][ContractKit]Add ability to do local signing

ashishb opened the pull request:
### Description

1. Add the ability to do a local signing
2. Add ability to send signed transactions
3. Stop `Web3ProviderEngine` after the test, instead of `forceExit`
4. Minor code cleanup

### Tested

`jest src/utils/tx-signing.test.ts` 

### Related issues

- Fixed #186 
- Fixes #477
- Fixed #811

mcortesi commented on packages/contractkit/src/utils/web3-utils.ts:
```diff
@@ -13,7 +13,7 @@ import { CeloProvider } from './tx-signing'
 
 // This web3 has signing ability.
-export async function addLocalAccount(web3: Web3, privateKey: string): Promise<Web3> {
+export function addLocalAccount(web3: Web3, privateKey: string): Web3 {
```
nice.. you removed the async

mcortesi commented on packages/contractkit/src/kit.ts:
```diff
@@ -67,6 +73,38 @@ export class ContractKit {
     return toTxResult(promiEvent)
   }
 
+  async sendSignedTransaction(tx: Tx): Promise<TransactionResult> {
```
i don't like this.
IMO, tx signing should be transparent, not something that you "say explicitly". You are already saying that when you use `from` with a local account

ashishb replied:
> You are already saying that when you use from with a local account

Not exactly true, as of now, someone might still want to have an account in both web3 as well as a locally running geth node and would want to sign to using the geth node. I can merge these two into one but I am sure someone later is going to demand the other use-case.

mcortesi replied:
why would something want like.. I think that's kind of a contrived example.

One should be able to do:

```ts
const stableToken = kit.contract.getStableToken()

const tx = await stableToken.transfer(value, to).send()

```

and that `send()` use a local account or not, depending on the `from`. which in this example is the default, but you can also do `.send({from: someAccount})`

ashishb replied:
I removed `sendSignedTransaction` completely. It should not be required at all (as confirmed by the unittest ts-signing.test.ts as well)

mcortesi commented on packages/contractkit/src/kit.ts:
```diff
@@ -67,6 +73,38 @@ export class ContractKit {
   }
 
+  async sendSignedTransaction(tx: Tx): Promise<TransactionResult> {
+    const gasFeeRecipient = await this.web3.eth.getCoinbase()
```
this should happen on the provider, not at `kit` level.
so, if i use web3 directly, i can still send a tx

mcortesi replied:
forget about this.... can happen either place! :P

ashishb replied:
Added them to `signTransactionAsync`. Also, removed `sendSignedTransaction` and added `gasFeeRecipient` calculation to `sendTransaction`. This makes that method async though.

mcortesi commented on packages/contractkit/src/test-utils/ganache-test.ts:
```diff
@@ -1,10 +1,20 @@
 export function jsonRpcCall<O>(web3: Web3, method: string, params: any[]): Promise<O> {
   return new Promise<O>((resolve, reject) => {
-    web3.currentProvider.send(
+    let currentProvider = web3.currentProvider
```
this shouldn't be addressed here....
for `jsonRpcCall` a provider is always a provider and all provider follow the same interface :P

ashishb replied:
Fair point. I moved this ugliness back into `tx-signing.test.ts`

mcortesi replied:
i still see the ugliness :P
```ts
if (currentProvider instanceof Web3ProviderEngine) {
      const providerEngine = web3.currentProvider as Web3ProviderEngine
      // @ts-ignore-next-line - I have to access private `_providers` variable
      currentProvider = providerEngine._providers.slice(-1)[0]._provider
    }
```

ashishb replied:
My bad. Removed.

mcortesi commented on packages/contractkit/src/utils/tx-signing.ts:
```diff
@@ -22,17 +29,65 @@ function getPrivateKeyWithout0xPrefix(privateKey: string) {
+  }
+
+  // Note: This method is async since it over-rides the parent async method.
+  public async getAccountsAsync(): Promise<string[]> {
```
why this?

ashishb replied:
Not required, I added it for brevity. Since I added a `sync` version. I am going to replace this with just a comment, how about that?

ashishb replied:
Also changes `getAccounts` to be `getAccount` why return a singleton array.

mcortesi commented on packages/contractkit/src/utils/tx-signing.ts:
```diff
@@ -22,17 +29,65 @@ function getPrivateKeyWithout0xPrefix(privateKey: string) {
+  ): Promise<void> {
+    let signingRequired = false
+    switch (payload.method) {
+      case 'eth_sendTransaction': // fallthrough
```
style: do `signingRequired = ['a','b','c'...].includes(method)`

ashishb replied:
Done

mcortesi commented on packages/contractkit/src/utils/web3-utils.ts:
```diff
@@ -12,24 +12,95 @@ import { CeloProvider } from './tx-signing'
+  return web3
+}
+
+function addLocalAccountToExistingProvider(
```
I think we can do this whole design better...

We can have:
`web3.currentProvider instanceOf CeloProvider`

And then create a CeloProvider, that internally proxies everything to the ProviderEngine, has a reference to the Transport/original provider. Knows the accounts, has a `addAccount()` method... etc.

mcortesi replied:
the current code seems unnecessarily complex

ashishb replied:
Good point. I am changing `CeloProvider` to `CeloPrivateKeyWalletProvider` and adding a new `CeloProvider` which hides all this complexity.

mcortesi commented on packages/contractkit/src/kit.ts:
```diff
@@ -49,24 +51,47 @@ export class ContractKit {
 
-  sendTransaction(tx: Tx): TransactionResult {
+  async sendTransaction(tx: Tx): Promise<TransactionResult> {
+    const gasFeeRecipient = await this.web3.eth.getCoinbase()
```
we should avoid doing all this work if the tx is not locally signed. since the node will on that case fill the gap. Neither should we fail if it's not a locally signed tx

ashishb replied:
This code has been removed.

mcortesi commented on packages/contractkit/src/kit.ts:
```diff
@@ -49,24 +51,47 @@ export class ContractKit {
     return toTxResult(promiEvent)
   }
 
+  async getGasPrice(gasCurrency: string | undefined): Promise<string | undefined> {
```
use `getGasPrice(gasCurrency?: string)` instead of using `undefined`

ashishb replied:
As discussed. This code has been removed.

mcortesi commented on packages/contractkit/src/kit.ts:
```diff
@@ -49,24 +51,47 @@ export class ContractKit {
     return toTxResult(promiEvent)
   }
 
+  async getGasPrice(gasCurrency: string | undefined): Promise<string | undefined> {
```
Add a jsdoc stating what it means to use gasCurrency parameter or not

ashishb replied:
As discussed. This code has been removed.

mcortesi commented on packages/contractkit/src/kit.ts:
```diff
@@ -49,24 +51,47 @@ export class ContractKit {
+  async getGasPrice(gasCurrency: string | undefined): Promise<string | undefined> {
+    // Gold Token
+    if (gasCurrency === undefined) {
+      return String(await this.web3.eth.getGasPrice())
```
why the `String()`?

ashishb replied:
As discussed. This code has been removed.

mcortesi commented on packages/contractkit/src/kit.ts:
```diff
@@ -49,24 +51,47 @@ export class ContractKit {
     return toTxResult(promiEvent)
   }
 
+  async getGasPrice(gasCurrency: string | undefined): Promise<string | undefined> {
```
when is `undefined` a valid return type???

ashishb replied:
As discussed. This code has been removed.

mcortesi commented on packages/contractkit/src/utils/tx-signing.ts:
```diff
@@ -20,38 +27,144 @@ function getPrivateKeyWithout0xPrefix(privateKey: string) {
 }
 
-export class CeloProvider extends PrivateKeyWalletSubprovider {
+function isNullOrUndefined(value: any): boolean {
```
style: unnecessary. In js, we use `x == null` to test this. In fact, that's the only valid use of `==` vs `===`.

mcortesi commented on packages/contractkit/src/utils/tx-signing.ts:
```diff
@@ -20,38 +27,144 @@ function getPrivateKeyWithout0xPrefix(privateKey: string) {
+    // Either signing is not required or
+    // signing is required and this class is the correct one to sign
+    const shouldPassToSuperClassForHandling =
+      !signingRequired || payload.params[0].from.toLowerCase() === this.accountAddress
```
utils/address has as `eqAddress()`

mcortesi commented on packages/contractkit/src/utils/tx-signing.ts:
```diff
@@ -20,38 +27,144 @@ function getPrivateKeyWithout0xPrefix(privateKey: string) {
   public async signTransactionAsync(txParams: CeloPartialTxParams): Promise<string> {
     debug('signTransactionAsync: txParams are %o', txParams)
-    if (txParams.chainId === undefined || txParams.chainId === null) {
+    if (txParams.from.toLowerCase() !== this.accountAddress) {
```
use `eqAddress`

mcortesi commented on packages/contractkit/src/utils/tx-signing.ts:
```diff
@@ -20,38 +27,144 @@ function getPrivateKeyWithout0xPrefix(privateKey: string) {
+      this.chainId = parseInt(result.result.toString(), 10)
+      debug('getChainId chain result ID is %s', this.chainId)
+    }
+    return this.chainId!
```
maybe check chainId != null and fail, instead of using `!`

mcortesi commented on packages/contractkit/src/utils/web3-utils.ts:
```diff
@@ -2,32 +2,32 @@ import {
+// Return the modified web3 object for chaining
+export function addLocalAccount(web3: Web3, privateKey: string): Web3 {
+  const existingProvider = web3.currentProvider
+  const providerEngine: CeloProvider =
```
change the var name to `celoProvider`?

ashishb replied:
Done.

mcortesi commented on packages/contractkit/src/utils/web3-utils.ts:
```diff
@@ -71,3 +71,68 @@ class WrappingSubprovider extends Subprovider {
+  private signingAddresses: string[]
+
+  constructor(readonly existingProvider: Provider) {
+    this.web3ProviderEngine = CeloProvider.createNewProviderWithLocalAccount(existingProvider)
```
maybe it's me, style: a bit weird to call a private static method from a constructor.

ashishb replied:
Removed.

mcortesi commented on packages/contractkit/src/utils/web3-utils.ts:
```diff
@@ -71,3 +71,68 @@ class WrappingSubprovider extends Subprovider {
+      return this
+    }
+
+    const localAccountProvider = new CeloPrivateKeyWalletProvider(privateKey)
```
imo, a better strategy to is have just ONE `CeloPrivateKeyWalletProvider` that handles many accounts. There a couple of things/state that they share (liked cached values for chainId and others).

This also avoids modifying the providerEngine providers list, mantain a list of providers, etc...

In fact, we can even store the accounts in the CeloProvider and pass a `getAccounts()` fn to the CeloPrivateKeyWalletProvider

ashishb replied:
Done. Changed CeloPrivateKeyWalletProvider to CeloPrivateKeysWalletProvider and this will take care of signing with the right key or doing nothing if it does not have the key.

mcortesi commented on packages/contractkit/src/utils/web3-utils.ts:
```diff
@@ -71,3 +71,68 @@ class WrappingSubprovider extends Subprovider {
+    this.web3ProviderEngine.sendAsync(payload, callback)
+  }
+
+  start(callback?: () => void): void {
```
let's turn this into a promise

ashishb replied:
Why?
I am just writing a delegate, so, why not keep the signature same as `web3Provider.start` for now?

mcortesi commented on packages/contractkit/src/utils/web3-utils.ts:
```diff
@@ -71,3 +71,68 @@ class WrappingSubprovider extends Subprovider {
+    this.web3ProviderEngine.on(event, handler)
+  }
+
+  send(payload: JSONRPCRequestPayload): void {
```
don't we miss a callback/promise here? how do i get the response?

ashishb replied:
I am just writing a delegate, so, the signature will match `web3ProviderEngine.send`

mcortesi replied:
solved. must match sendAsync for web3 to work

mcortesi commented on packages/contractkit/src/utils/web3-utils.ts:
```diff
@@ -71,3 +71,68 @@ class WrappingSubprovider extends Subprovider {
+    return this
+  }
+
+  on(event: string, handler: () => void): void {
```
we should be smarter here.
IPC and WS provider have `on`, and thus "support subscriptions" in web3js terms, while HTTP doesn't. If we set this provider, web3js will understand that WE support subscriptions always

ashishb replied:
Good catch. I didn't know that. As discussed changed to initializing it conditionally. One minor issue is that I  couldn't figure out the type of `on`, so, it is

`public readonly on: null | any`

I tried `public readonly on: null | (string, () => void)` but that doesn't seem to work.

All other delegates are created unconditionally.

mcortesi replied:
solved

ashishb committed:
Add ability to do local signing

Description
1. Add the ability to do a local signing
2. Add ability to send signed transactions
3. Stop Web3ProviderEngine after the test, instead of forceExit
4. Minor code cleanup

Tested
jest src/utils/tx-signing.test.ts

mcortesi committed:
cleanup, fix send(), typings

mcortesi commented:
I think what's missing is digging deeper into providerEngine start() and what it actually does.

Also, we need to manage disconnection for the wrapping subprovider and do it when we do `stop`.
In fact, we can do that in `CeloProvider.stop()` impl,

ashishb committed:
Close underlying providers

ashishb commented:
> Also, we need to manage disconnection for the wrapping subprovider and do it when we do stop. In fact, we can do that in CeloProvider.stop() impl,

Added
```
    try {
      if (this.underlyingProvider.hasOwnProperty('stop')) {
        // @ts-ignore
        this.underlyingProvider.stop()
      }
  
      // Close the web3 connection or the CLI hangs forever.
      if (this.underlyingProvider && this.underlyingProvider.hasOwnProperty('connection')) {
        // @ts-ignore
        const connection = this.underlyingProvider.connection
        if (connection.hasOwnProperty('_connection')) {
          connection._connection.close()
        }
      }
    } catch (error) {
      debug(`Failed to close the connection: ${error}`)
    }
```

ashishb commented:
> I think what's missing is digging deeper into providerEngine start() and what it actually does.

It seems to me that it is for polling block afterwards to primarily continuously track the latest block. I think I can disable it by passing `opts` with `blockTracker` object. https://github.com/MetaMask/web3-provider-engine/blob/master/index.js#L26

mcortesi commented:
> > I think what's missing is digging deeper into providerEngine start() and what it actually does.
> 
> It seems to me that it is for polling block afterwards to primarily continuously track the latest block. I think I can disable it by passing `opts` with `blockTracker` object. https://github.com/MetaMask/web3-provider-engine/blob/master/index.js#L26

what about completely replacing ProviderEngine?

For what i see, there not much going one there... it emits a couple of event nobody cares about, it add this start/stop behaviour that we don't like. And the subprovider that we use only needs to call `sendAsync()`

mcortesi commented on packages/contractkit/src/providers/celo-private-keys-subprovider.ts:
```diff
@@ -0,0 +1,192 @@
+
+  public addAccount(privateKey: string) {
+    // Prefix 0x here or else the signed transaction produces dramatically different signer!!!
+    privateKey = '0x' + getPrivateKeyWithout0xPrefix(privateKey)
```
celo/utils has or should have an `ensure0x` (if it's not in utils it's celotool and should be ported)

ashishb replied:
It is in celotool and it is broken `export const ensure0x = (hexstr: string) => (hexstr.startsWith('0x') ? hexstr : '0x' + hexstr)` since it will miss a string which starts with `0X`. The consquence of that would be a disaster. Let's fix that in a separate PR, so that, 5 unrelated files are not impacted.

```
 $ rg ensure0x
celotool/src/lib/generate_utils.ts
12:import { ensure0x, strip0x } from '@celo/celotool/src/lib/utils'
77:  return new Web3.modules.Eth().accounts.privateKeyToAccount(ensure0x(privateKey)).address

celotool/src/lib/utils.ts
146:export const ensure0x = (hexstr: string) => (hexstr.startsWith('0x') ? hexstr : '0x' + hexstr)

celotool/geth_tests/src/lib/utils.ts
11:import { ensure0x } from '@celo/celotool/src/lib/utils'
333:    validatorPrivateKeys.map(ensure0x).join(','),

celotool/src/cmds/geth/get_gold_balance.ts
5:  ensure0x,
41:  account = ensure0x(account)

celotool/src/cmds/geth/run.ts
5:import { addCeloGethMiddleware, ensure0x, validateAccountAddress } from 'src/lib/utils'
80:      (minerAddress: string) => (minerAddress === null ? null : ensure0x(minerAddress))

celotool/src/cmds/deploy/initial/contracts.ts
6:import { ensure0x, execCmd } from '@celo/celotool/src/lib/utils'
24:    return ensure0x(OG_ACCOUNTS[0].address)
34:    return OG_ACCOUNTS.map((account) => account.privateKey).map(ensure0x)
40:    ).map(ensure0x)
```

mcortesi replied:
sure, let's do that! do you want to create and issue or do that after this pr?

ashishb replied:
I will send a follow-on PR to fix `ensure0x`. Let's not worry about de-duplicating this function for now.

mcortesi commented on packages/contractkit/src/utils/tx-signing.test.ts:
```diff
@@ -1,56 +1,81 @@
-    it('should be able to sign and get the signer back', async () => {
+  let originalProvider: Provider
+  beforeEach(() => {
+    originalProvider = web3.currentProvider
```
I'm a bit lost here.. but I'm missing the part were we set the web3.currentProvider to CeloProvider. Where does that happen?

ashishb replied:
see `addLocalAccount` in web3-utils.ts - https://github.com/celo-org/celo-monorepo/pull/798/files#diff-728b956ec7d19ed74a7cb0a6beb7be98

We change to CeloProvider when the first local account is added.

mcortesi commented:
@ashishb approved. Great Job.

I think what's missing is:
 - disable block tracker on provider engine

then we can do some improvements, but we can do that in another time ;)
